core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyContinuousInValues.java

```diff
@@ -0,0 +1,127 @@
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.sql.ir.IrUtils.or;
+
+/**
+ * Simplify IN expression with continuous range of constant test values into a BETWEEN expression. E.g,
+ * <ul>
+ *     <li>{@code $in(x, [1, 2, 3, 4]) -> $between(x, 1, 4)}
+ * </ul>
+ */
+public class SimplifyContinuousInValues
+        implements IrOptimizerRule
+{
+    @Override
+    public Optional<Expression> apply(Expression expression, Session session, Map<Symbol, Expression> bindings)
+    {
+        if (!(expression instanceof In in)) {
+            return Optional.empty();
+        }
+
+        return simplifyContinuousInValues(in);
+    }
+
+    public static Optional<Expression> simplifyContinuousInValues(In in)
+    {
+        if (in.valueList().size() < 2) {
+            return Optional.empty();
+        }
+
+        Type valueType = in.value().type();
+        if (!isIntegerArithmeticValidType(valueType)) {
+            return Optional.empty();
+        }
+
+        boolean nullMatch = false;
+        long nonNullsCount = 0;
+        long min = Long.MAX_VALUE;
+        long max = Long.MIN_VALUE;
+        for (Expression testExpression : in.valueList()) {
+            if (!(testExpression instanceof Constant constant)) {
+                return Optional.empty();
+            }
+            if (constant.value() == null) {
+                nullMatch = true;
+                continue;
+            }
+            long longConstant = ((Number) constant.value()).longValue();
+            min = Math.min(min, longConstant);
+            max = Math.max(max, longConstant);
+            nonNullsCount++;
+        }
+
+        // If all values within a range are included, use a range filter
+        if (nonNullsCount >= 2 && (max - min + 1) == nonNullsCount) {
+            Between between = new Between(in.value(), new Constant(valueType, min), new Constant(valueType, max));
+            if (nullMatch) {
+                return Optional.of(or(new IsNull(in.value()), between));
+            }
+            return Optional.of(between);
+        }
+
+        return Optional.empty();
+    }
+
+    private static boolean isIntegerArithmeticValidType(Type type)
+    {
+        if (type.getJavaType() != long.class) {
+            return false;
+        }
+        // Types for which we can safely use equality and comparison on the stored long value
+        // instead of going through type specific methods
+        return type == TINYINT ||
+                type == SMALLINT ||
+                type == INTEGER ||
+                type == BIGINT ||
+                type == TIME_SECONDS ||
+                type == TIME_MILLIS ||
+                type == TIME_MICROS ||
+                type == TIME_NANOS ||
+                type == TIME_PICOS ||
+                type == DATE ||
+                type == TIMESTAMP_SECONDS ||
+                type == TIMESTAMP_MILLIS ||
+                type == TIMESTAMP_MICROS ||
```

Comment:
What about time and timestamp with other precisions? Also, test with instanceof, as there's no guarantee the instances are the same.

Reply:
updated this